drivers/mailbox/bflb-ipc.c

```diff
@@ -157,17 +192,42 @@ static const struct irq_domain_ops bflb_ipc_irq_ops = {
 	.xlate = bflb_ipc_domain_xlate,
 };
 
+/* JH: Figure out if M0 has processed the last signal sent
+ * by checking if the High/Low registers are cleared
+ */
```

Comment:
We probably want an ACK interrupt from M0 for this, otherwise it's a race with the clearing of the registers.

Reply:
Ok, then we would need two mailboxes (one for sending, one for recieving) right?

also, should I move the ```mboxes = <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_MBOX_VIRTIO_KICK>;``` thats set in the rproc device to the ipclic instead?

Reply:
There's one set of the ILSLR/ILSHR pair for communication M0->D0 at ipc->base[1] and another set for communication D0->M0 at ipc->base[2].  I don't know if we really need the the M0->D0 mailboxes though, since we can only have one mailbox message outstanding at once unless some memory based ring buffer is added, it's always known which mbox write is being ACKed when the mbox interrupt comes back.

I like the idea of having a mailbox dt cell that donates the CPU target like you did, I was imagining expanding mbox-cells to 3 and splitting the last parameter to <client> and <signal>.  That way we could reuse the client numbers from ./include/dt-bindings/mailbox/bflb-ipc.h.  The exact format of what goes into the cells is arbitrary and is just an agreement between the mailbox client (rproc) and mailbox provider (ipclic).  The mboxes entry should be associated with the mailbox client (rproc) and is way to identify what the mailbox subsystem calls a "channel" and specifies the arguments that will get sent to the of_xlate() callback of the mailbox provider.  If rproc only requires a single channel to M0 then it would just have one mbox entry like this:

`mboxes = <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_KICK>;`

Then in your remoteproc driver:
`chan = mbox_request_channel(&mbox_client, 0);`


If you need multiple channels, use mbox-names and allocate the channels by friendly name:
```
mboxes = <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_ALLOC>,
         <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_PING>,
         <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_KICK>;
mbox-names = "alloc", "ping", "kick";
```
...
```
chan1 = mbox_request_channel_byname(&mbox_client, "alloc");
chan2 = mbox_request_channel_byname(&mbox_client, "ping");
chan3 = mbox_request_channel_byname(&mbox_client, "kick");
```

Reply:
Ok, few more questions before I dive in and update:


> There's one set of the ILSLR/ILSHR pair for communication M0->D0 at ipc->base[1] and another set for communication D0->M0 at ipc->base[2]. I don't know if we really need the the M0->D0 mailboxes though, since we can only have one mailbox message outstanding at once unless some memory based ring buffer is added, it's always known which mbox write is being ACKed when the mbox interrupt comes back.
> 

I think i need both ILSLR/ILSHR registers as both sides need to signal
1) there is new data in the TX ring (from M0 or D0)
2) I have consumed the data in a RX ring (from M0 or D0)

So I can use the BFLB_IPC_TARGET_* as the index to the ipc->base register right?


> I like the idea of having a mailbox dt cell that donates the CPU target like you did, I was imagining expanding mbox-cells to 3 and splitting the last parameter to "client" and "signal". That way we could reuse the client numbers from ./include/dt-bindings/mailbox/bflb-ipc.h. 

Ok, but just for clarity - the remoteproc sends a index of what ring its "kicking" on the other side. in reality its only 0 or 1, but the core driver "allocates" a u32 as the index... so:
pack client/signal into say ILSHR (16 bits each) and the "index" into ILSHR?


>The exact format of what goes into the cells is arbitrary and is just an agreement between the mailbox client (rproc) and mailbox provider (ipclic). The mboxes entry should be associated with the mailbox client (rproc) and is way to identify what the mailbox subsystem calls a "channel" and specifies the arguments that will get sent to the of_xlate() callback of the mailbox provider. If rproc only requires a single channel to M0 then it would just have one mbox entry like this:
> 
> `mboxes = <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_KICK>;`
> 
> Then in your remoteproc driver: `chan = mbox_request_channel(&mbox_client, 0);`
> 
> If you need multiple channels, use mbox-names and allocate the channels by friendly name:
> 
> ```
> mboxes = <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_ALLOC>,
>          <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_PING>,
>          <&ipclic BFLB_IPC_TARGET_M0 BFLB_IPC_DEVICE_RPROC BFLB_IPC_MBOX_VIRTIO_KICK>;
> mbox-names = "alloc", "ping", "kick";
> ```
> 
> ...
> 
> ```
> chan1 = mbox_request_channel_byname(&mbox_client, "alloc");
> chan2 = mbox_request_channel_byname(&mbox_client, "ping");
> chan3 = mbox_request_channel_byname(&mbox_client, "kick");
> ```

Got it.
and just to confirm... nothing stopping the mailbox driver being bi-directional right?